plugin_ide.ui/src/com/github/rustdt/ide/ui/editor/RustEditorActionContributor.java

```diff
@@ -28,6 +51,99 @@ protected RustOpenDefinitionOperation createOpenDefinitionOperation(ITextEditor
 	
 	@Override
 	protected void registerOtherEditorHandlers() {
+		setupOpenType();
+	}
+	
+	private void setupOpenType() {
+		evaluateGlobalSourceStructure();
+		listenToUpdatesOfGlobalSourceStructure();
+	}
+	
+	private void evaluateGlobalSourceStructure() {
+		new Job("Searching types") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					ResourcesPlugin.getWorkspace().getRoot().accept(RustEditorActionContributor.this::addResource);
+				} catch(Exception e) {
+					AbstractLangCore.log().logError("Global source structure could not be determined", e);
+				}
+				return Status.OK_STATUS;
+			}
+		}.schedule();
+	}
+	
+	private boolean addResource(IResource resource) {
+		convertToRustFileLocation(resource).ifPresent(RustEditorActionContributor::fileTouched);
+		return true;
+	}
+	
+	private void listenToUpdatesOfGlobalSourceStructure() {
+		ResourcesPlugin.getWorkspace().addResourceChangeListener(this::processEvent, IResourceChangeEvent.POST_CHANGE);
 	}
 	
+	private void processEvent(IResourceChangeEvent event) {
+		new Job("Updating types") {
+			@Override
+			protected IStatus run(IProgressMonitor monitor) {
+				try {
+					event.getDelta().accept(RustEditorActionContributor.this::applyResourceDelta);
+				} catch(Exception e) {
+					AbstractLangCore.log().logError("Global source structure could not be determined", e);
+				}
+				return Status.OK_STATUS;
+			}
+		}.schedule();
+	}
+	
+	private boolean applyResourceDelta(IResourceDelta delta) {
+		convertToRustFileLocation(delta.getResource()).ifPresent(location -> {
+			switch(delta.getKind()) {
+			case IResourceDelta.ADDED:
+			case IResourceDelta.CHANGED:
+				fileTouched(location);
+				break;
+			case IResourceDelta.REMOVED:
+				fileRemoved(location);
+				break;
+			}

```

Comment:
For this to have correct concurrent behavior, you will need to analyze the delta in the same thread that triggers the delta. However, you cannot analyze the file itself in the delta thread, as that is too slow. What needs to be done is that GlobalSourceStructure (or a similar class) must take not immediately (in the delta thread) of which files have been modified/removed, and then queue a source structure update in some background thread. (Possibly cancelling the previous update if it is still running).  You can see an example of this in DocumentReconcileManager, which is responsible for listening to document changes (keypresses), and notifying the SourceModelManager that a new update is needed.

Also, this way the GlobalSourceStructure always knows which files are updated, and which are not (this can then be used in a UI display like the Open-Type dialog, where it can display a text like "Indexing 23 files...", in a similar way to what JDT does)